### ISSUE TITLE
Setting seed for population creation

### ISSUE THREAD

**Issue by PavelBal** [Open-Systems-Pharmacology/PK-Sim#1634]
For reproducibility, it should be possible to set the seed used by the PRNG to create populations.

**Comment by msevestre** [comment]
What's PRNG?

**Comment by PavelBal** [comment]
Pseudo-random number generator ;)

**Comment by Yuri05** [comment]
>Pseudo-random number generator ;)

And I thought it would be [this](https://www.acronymfinder.com/Penicillin-Resistant-Neisseria-Gonorrhea-(PRNG).html) :rofl:

**Comment by PavelBal** [comment]
> And I thought it would be this

I could answer "у кого что болит...", but I better not :D

**Comment by msevestre** [comment]
This is already implemented in the Snapshot. I am not sure why you would need something like this in PKSim

**Comment by PavelBal** [comment]
As mentioned above - to have reproducible results. Also outside of snapshot re-creation. E.g. - two populations that should be identical except for one parameter with user defined variability.

If the functionality is already present - why not expose it as a text field that can be activated/deactivated?

**Comment by StephanSchaller** [comment]
I do not think it will confuse users, and I also believe this is an important feature for better reproducibility.

**Comment by msevestre** [comment]
Let's agree to disagree. 
This can be done at the moment via snapshot and in R.

**Comment by Yuri05** [comment]
I think if we enable saving/loading of Building Blocks to/from snapshot (#1312) - it would be enough? @PavelBal 
(which can be done already now in dev mode)

**Comment by PavelBal** [comment]
> it would be enough?

I still think that having seed as an advanced parameter is a better solution. This way we could also ensure that populations created in PK-Sim and in R are identical (which will not be possible with snapshots only).

**Comment by Yuri05** [comment]
Still, I would not provide the seed as an **input** parameter in PK-Sim.
I would at most show the used seed value after the population was created. Then you could use it in R etc.

Right now, if you need to recreate the same population in R: save the project to snapshot (or save just the population to snapshot in dev mode) and take the seed from the saved snapshot.

By the way: the population seed is in general not enough. You need additional seed for every user defined parameter distribution:

![grafik](https://user-images.githubusercontent.com/25061876/202176355-a66f7471-0506-4d5a-bd96-b89a7dbf95e5.png)

**Comment by PavelBal** [comment]
> By the way: the population seed is in general not enough. You need additional seed for every user defined parameter distribution:

Wow! Why? Why is the population seed not enough? This makes the results nearly not reproducible.... And I honestly cannot understand why there is a separate seed for each parameter?

**Comment by Yuri05** [comment]
Because advanced parameters are created separately from the "main" population creation.

Of course you could use the population seed for each advanced parameter. But then if you e.g. have 2 advanced parameters with the same distribution properties (e.g. "`Normal distribution; Mean=1; SD=0.5`") - all those parameters would become exactly the same values in the population - which is not what you want.

In the "main" population generation it is different, because the random generator is initialized once at the beginning (with the population seed) and is then sequentially used for all distributed population parameters.

**Comment by PavelBal** [comment]
> In the "main" population generation it is different, because the random generator is initialized once at the beginning (with the population seed) and is then sequentially used for all distributed population parameters.

Why not re-use the same RNG instance also for advanced parameters?

**Comment by Yuri05** [comment]
>Why not re-use the same RNG instance also for advanced parameters? 

Theoretically possible. But then you also need to keep track **in which order advanced parameters were added**

**Comment by Yuri05** [comment]
More than that: you can remove advanced parameters again. So if you want to keep the same Random generator and use the same seed - you also need to keep track of every Add/Remove operation, which would be totally crazy.

**Comment by Yuri05** [comment]
So in short: I don't see an easy possibility to use the population seed for reproducible recreation of advanced parameters.
But maybe @msevestre has an idea

**Comment by PavelBal** [comment]
> possibility to use the population seed for reproducible recreation of advanced parameters.

Use the global population seed to generate seeds of advanced parameters?

**Comment by PavelBal** [comment]
> Well, we could do something funky like

I like the idea!
